sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/DataSourceV2Relation.scala

```diff
@@ -224,24 +226,29 @@ object DataSourceV2Relation {
         )
 
       case filterSupport: SupportsPushDownFilters =>
-        // A map from original Catalyst expressions to corresponding translated data source
-        // filters. If a predicate is not in this map, it means it cannot be pushed down.
-        val translatedMap: Map[Expression, Filter] = filters.flatMap { p =>
-          DataSourceStrategy.translateFilter(p).map(f => p -> f)
-        }.toMap
+        val translateResults: Seq[(Option[Filter], Expression)] = filters.map { p =>
+          (DataSourceStrategy.translateFilter(p), p)
+        }
 
         // Catalyst predicate expressions that cannot be converted to data source filters.
-        val nonConvertiblePredicates = filters.filterNot(translatedMap.contains)
+        val nonConvertiblePredicates = translateResults.collect {
```

Comment:
Avoid using the `predicates` and `filters` as the variable names in the same function. This just makes the reader confused.